spec/controllers/registrations_controller_spec.rb

```diff
@@ -14,35 +14,157 @@
+
+        user = User.last
+        expect(user.email).to eql('foo@example.com')
+        expect(user.confirmed_at).to be_nil
+        expect(ActionMailer::Base.deliveries.count).to eql(1)
+        expect(ActionMailer::Base.deliveries.first.subject).to eql("[Shuttle] Confirmation instructions")
+      end
+    end
+
+    context 'with bogus email' do
+      let(:email) { 'foo123' }
+      let(:extra_params) { { 'REMOTE_ADDR' => '101.0.0.2' } }
+
+      it 'should handle bogus email addresses' do
+        response = subject
+        expect(response).to eq(200)
+
+        expect(User.count).to be_zero
+      end
+    end
+
+    context 'throttling requests' do
+      let(:email) { 'foo@example.com' }
+      let(:extra_params) { { 'REMOTE_ADDR' => '101.0.0.3' } }
+
+      it 'fails after 20 tries' do
+        20.times.each do |index|
+          params[:user][:email] = "foo-#{index}@example.com"
+          response = post(url, params, extra_params)
+          expect(response).to eq(302)
+        end
+        expect(User.count).to eq(20)
+
+        params[:user][:email] = "foo-21@example.com"
+        response = post(url, params, extra_params)
+        expect(response).to eq(503)
+        expect(User.count).to eq(20)
+      end
+    end
+  end
+
+  describe '#sign_in' do
+    let(:url) { 'http://test.host/users/sign_in' }
+    let(:email) { 'foo@example.com' }
+    let(:password) { 'test123'}
+    let(:user) do
+      {
+        email:    email,
+        password: password,
       }
-      user = User.last
-      expect(user.email).to eql('foo@example.com')
-      expect(user.confirmed_at).to be_nil
-      expect(ActionMailer::Base.deliveries.count).to eql(1)
-      expect(ActionMailer::Base.deliveries.first.subject).to eql("[Shuttle] Confirmation instructions")
     end
 
-    it "should handle bogus email addresses" do
-      post :create, user: {
-          email:                 'foo123',
+    before do
+      login_params = {
+        user: {
+          email:                 email,
           first_name:            'Sancho',
           last_name:             'Sample',
-          password:              'test123',
-          password_confirmation: 'test123'
+          password:              password,
+          password_confirmation: password
+        }
       }
-      expect(User.count).to be_zero
+      post 'http://test.host/users', login_params
+    end
+
+    context 'with proper credential' do
+      let(:extra_params) { { 'REMOTE_ADDR' => '102.0.0.1' } }
+
+      it 'logged in successfully' do
+        response = subject
+        expect(response).to eq(302)
+      end
+    end
+
+    context 'with invalid credential' do
+      let(:email) { 'foo123' }
+      let(:extra_params) { { 'REMOTE_ADDR' => '102.0.0.2' } }
+
+      it 'failed to login' do
+        response = subject
+        expect(response).to eq(200)
+      end
+    end
+
+    context 'throttling requests' do
+      let(:extra_params) { { 'REMOTE_ADDR' => '102.0.0.3' } }
```

Comment:
We should also check that someone with a different IP can still proceed even if someone else is throttled.